From bbba6c60bad843674e8de54380dc6d558805917d Mon Sep 17 00:00:00 2001
From: Samuel Dowling <samuel.dowling@protonmail.com>
Date: Thu, 30 Sep 2021 22:38:20 +0930
Subject: [PATCH] Guard dependency discovery

* Add guards to prevent usage of custom cmake find package scripts.
* Export all symbols on windows
---
 CMakeLists.txt | 83 ++++++++++++++++++++++++++++++++++++++++++--------
 1 file changed, 70 insertions(+), 13 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index f646ca2..733185d 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -273,7 +273,13 @@ if(APPLE)
   set(ARMA_USE_ACCELERATE true)
   
   if(ALLOW_OPENBLAS_MACOS)
-    include(ARMA_FindOpenBLAS)
+    if(USE_SYSTEM_OPENBLAS)
+        include(ARMA_FindOpenBLAS)
+    elseif(USE_OPENBLAS)
+        find_package(OpenBLAS)
+    else()
+        set(OpenBLAS_FOUND NO)
+    endif()
     message(STATUS "OpenBLAS_FOUND = ${OpenBLAS_FOUND}")
     message(STATUS "")
     message(STATUS "*** If use of OpenBLAS is causing problems,")
@@ -288,8 +294,16 @@ if(APPLE)
   endif()
   
   if(ALLOW_BLAS_LAPACK_MACOS)
-    include(ARMA_FindBLAS)
-    include(ARMA_FindLAPACK)
+    if(USE_SYSTEM_BLAS)
+      include(ARMA_FindBLAS)
+    else()
+      set(BLAS_FOUND NO)
+    endif()
+    if(USE_SYSTEM_LAPACK)
+      include(ARMA_FindLAPACK)
+    else()
+      set(LAPACK_FOUND NO)
+    endif()
     message(STATUS "  BLAS_FOUND = ${BLAS_FOUND}"  )
     message(STATUS "LAPACK_FOUND = ${LAPACK_FOUND}")
     message(STATUS "")
@@ -328,14 +342,44 @@ if(APPLE)
   
 else()
   
-  include(ARMA_FindMKL)
-  include(ARMA_FindOpenBLAS)
-  include(ARMA_FindATLAS)
-  include(ARMA_FindBLAS)
-  include(ARMA_FindLAPACK)
+  if(USE_SYSTEM_MKL)
+    include(ARMA_FindMKL)
+  else()
+    set(MKL_FOUND NO)
+  endif()
+
+  if(USE_SYSTEM_OPENBLAS)
+    include(ARMA_FindOpenBLAS)
+  elseif(USE_OPENBLAS)
+    find_package(OpenBLAS)
+  else()
+    set(OpenBLAS_FOUND NO)
+  endif()
+
+  if(USE_SYSTEM_ATLAS)
+    include(ARMA_FindATLAS)
+  else()
+    set(ATLAS_FOUND NO)
+  endif()
+
+  if(USE_SYSTEM_BLAS)
+    include(ARMA_FindBLAS)
+  else()
+    set(BLAS_FOUND NO)
+  endif()
+
+  if(USE_SYSTEM_LAPACK)
+    include(ARMA_FindLAPACK)
+  else()
+    set(LAPACK_FOUND NO)
+  endif()
   
   if(ALLOW_FLEXIBLAS_LINUX AND (${CMAKE_SYSTEM_NAME} MATCHES "Linux"))
-    include(ARMA_FindFlexiBLAS)
+    if(USE_SYSTEM_FLEXIBLAS)
+      include(ARMA_FindFlexiBLAS)
+    else()
+      set(FlexiBLAS_FOUND NO)
+    endif()
   endif()
   
   message(STATUS "      MKL_FOUND = ${MKL_FOUND}"       )
@@ -469,7 +513,12 @@ if(DETECT_HDF5)
     # HDF5_INCLUDE_DIRS is the correct include directory.  So, in either case we
     # can use the first element in the list.  Issue a status message, too, just
     # for good measure.
-    list(GET HDF5_INCLUDE_DIRS 0 ARMA_HDF5_INCLUDE_DIR)
+    if(NOT USE_SYSTEM_HDF5)
+      list(GET HDF5_INCLUDE_DIRS 1 ARMA_HDF5_INCLUDE_DIR)
+    else()
+      list(GET HDF5_INCLUDE_DIRS 0 ARMA_HDF5_INCLUDE_DIR)
+    endif()
+
     message(STATUS "ARMA_HDF5_INCLUDE_DIR = ${ARMA_HDF5_INCLUDE_DIR}")
     message(STATUS "")
     message(STATUS "*** If use of HDF5 is causing problems,")
@@ -479,7 +528,11 @@ if(DETECT_HDF5)
   endif()
 endif()
 
-include(ARMA_FindARPACK)
+if(USE_SYSTEM_ARPACK)
+  include(ARMA_FindARPACK)
+else()
+  set(ARPACK_FOUND NO)
+endif()
 message(STATUS "ARPACK_FOUND = ${ARPACK_FOUND}")
 
 if(ARPACK_FOUND)
@@ -487,7 +540,11 @@ if(ARPACK_FOUND)
   set(ARMA_LIBS ${ARMA_LIBS} ${ARPACK_LIBRARY})
 endif()
 
-include(ARMA_FindSuperLU5)
+if(USE_SYSTEM_SUPERLU)
+  include(ARMA_FindSuperLU5)
+else()
+  set(SuperLU_FOUND NO)
+endif()
 message(STATUS "SuperLU_FOUND = ${SuperLU_FOUND}")
 
 if(SuperLU_FOUND)
@@ -551,7 +608,7 @@ message(STATUS "CMAKE_CXX_FLAGS           = ${CMAKE_CXX_FLAGS}"          )
 message(STATUS "CMAKE_SHARED_LINKER_FLAGS = ${CMAKE_SHARED_LINKER_FLAGS}")
 message(STATUS "CMAKE_REQUIRED_INCLUDES   = ${CMAKE_REQUIRED_INCLUDES}"  )
 
-
+set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
 add_library( armadillo ${PROJECT_SOURCE_DIR}/src/wrapper1.cpp ${PROJECT_SOURCE_DIR}/src/wrapper2.cpp )
 target_link_libraries( armadillo ${ARMA_LIBS} )
 # target_include_directories(armadillo INTERFACE $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include> $<INSTALL_INTERFACE:include>)
-- 
2.33.0

